### PR TITLE
Navigation: Add tab title to breadcrumbs

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- Adds the tab title to the breadcrumbs for tabs after the first

**Why do we need this feature?**

- clearer UX

**Who is this feature for?**

- everyone

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Review by eledobleefe — APPROVED** [review]
LGTM 👌

File: public/app/core/components/Breadcrumbs/utils.ts [hunk 1]
```diff
@@ -35,7 +35,16 @@ export function buildBreadcrumbs(sectionNav: NavModelItem, pageNav?: NavModelIte
     const shouldAddCrumb = !node.hideFromBreadcrumbs && !(shouldDedupe && isSamePathAsLastBreadcrumb);
 
     if (shouldAddCrumb) {
-      crumbs.unshift({ text: node.text, href: node.url ?? '' });
+      const activeChildIndex = node.children?.findIndex((child) => child.active) ?? -1;
```
Comment (Clarity-89) [inline public/app/core/components/Breadcrumbs/utils.ts:38]: Doesn't `findIndex` return `-1` by default if no elements found?

Comment (ashharrison90) [reply public/app/core/components/Breadcrumbs/utils.ts:38]: yes, but `node.children` could be `undefined`

**Review by joshhunt — COMMENTED** [review]
I reviewed this and it's... fine I guess.

I think showing the tab in brackets is kinda wonky tbh. I think it would be better if we presented tabs as a seperate breadcrumb... I know we've discussed this, but I can't remember why we avoided it?

**Comment by ashharrison90** [comment]
> I reviewed this and it's... fine I guess.
> 
> I think showing the tab in brackets is kinda wonky tbh. I think it would be better if we presented tabs as a seperate breadcrumb... I know we've discussed this, but I can't remember why we avoided it?

because then tab 1 becomes a "parent" of tab n, where n > 1. think about something like dashboard variables:

image

i think you're saying you'd want something like `Home -> Dashboards -> <dashboardName> -> Settings -> Variables`? but then where does `Settings` link to? the only thing that really makes sense is the `General` tab. so you have a weird situation where you've got 2 hierarchical breadcrumbs that link to sibling tabs 😕 

with this change, the breadcrumbs would be `Home -> Dashboards -> <dashboardName> -> Settings (Variables)`

**Comment by dprokop** [comment]
> i think you're saying you'd want something like Home -> Dashboards -> <dashboardName> -> Settings -> Variables? but then where does Settings link to? the only thing that really makes sense is the General tab. so you have a weird situation where you've got 2 hierarchical breadcrumbs that link to sibling tabs 😕

> with this change, the breadcrumbs would be Home -> Dashboards -> <dashboardName> -> Settings (Variables)

@ashharrison90 - what about skipping the `Settings` in dashboard breadcrumbs altogether, and instead just having the current active tab as a breadcrumb...? There is no intermediate Settings page anyways, so i think we can consider having dash settings breadcrumbs like this:

`Home -> Dashboards -> <dashboardName> -> General` or General settings
`Home -> Dashboards -> <dashboardName> -> Variables`
`Home -> Dashboards -> <dashboardName>  -> Annotations`

You get the point.

I agree with @joshhunt that the tab name in brackets is awkward.

**Comment by joshhunt** [comment]
> but then where does Settings link to? the only thing that really makes sense is the General tab. so you have a weird situation where you've got 2 hierarchical breadcrumbs that link to sibling tabs 😕

I understand and kind of agree @ashharrison90, but I think that's less-worse than the brackets.

But thinking about other apps that have much more sibling-y tabs, I think @dprokop's suggestion is better.

**Comment by ashharrison90** [comment]
> > i think you're saying you'd want something like Home -> Dashboards ->  -> Settings -> Variables? but then where does Settings link to? the only thing that really makes sense is the General tab. so you have a weird situation where you've got 2 hierarchical breadcrumbs that link to sibling tabs 😕
> 
> > with this change, the breadcrumbs would be Home -> Dashboards ->  -> Settings (Variables)
> 
> @ashharrison90 - what about skipping the `Settings` in dashboard breadcrumbs altogether, and instead just having the current active tab as a breadcrumb...? There is no intermediate Settings page anyways, so i think we can consider having dash settings breadcrumbs like this:
> 
> `Home -> Dashboards -> <dashboardName> -> General` or General settings `Home -> Dashboards -> <dashboardName> -> Variables` `Home -> Dashboards -> <dashboardName> -> Annotations`
> 
> You get the point.
> 
> I agree with @joshhunt that the tab name in brackets is awkward.

@dprokop @joshhunt that works _ok_ for dashboard settings, but doesn't work at all for apps with tabs at the root. you end up in a situation where you have `Home -> Apps -> <tabName>` and lose the app name completely. 

gonna list the solutions i think are currently feasible, feel free to suggest others:
- this one
  - for dashboard settings, this would be `Home -> Dashboards -> <dashboardName> -> Settings (Variables)`
  - for an app with root tabs, this would be `Home -> Apps -> <appName> (<tabName>)`
- tabs as a completely separate breadcrumb
  - for dashboard settings, this would be `Home -> Dashboards -> <dashboardName> -> Settings -> Variables` where `Settings` links to the `General` tab
  - for an app with root tabs, this would be `Home -> Apps -> <appName> -> <tabName>` where `<appName>` links to the first tab

in both cases, i guess we wouldn't show the child breadcrumb tab if it was the first tab, so it would be `Home -> Apps -> <appName>` for the first tab
    
my vote would be the solution in this PR, since it doesn't create a situation where sibling tabs are listed as parents in the breadcrumbs. but i can see the argument for the other one as well. 

any other suggestions? otherwise cast your vote and i will bow to democracy! 🙇

**Comment by joshhunt** [comment]
@ashharrison90 

Is it possible to address that shortcoming specifically? Can we enforce that an App always has a breadcrumb? The way I think about breadcrumbs is that it's actually structured something like this

```
 Grafana/nav        App              App Pages
[Home > Apps] > [Acme Corp] > [Metrics > Pod restarts]
```

I guess this aligns to your second proposal. I think that's much better.

**Comment by ashharrison90** [comment]
> @ashharrison90
> 
> Is it possible to address that shortcoming specifically? Can we enforce that an App always has a breadcrumb? The way I think about breadcrumbs is that it's actually structured something like this
> 
> ```
>  Grafana/nav        App              App Pages
> [Home > Apps] > [Acme Corp] > [Metrics > Pod restarts]
> ```
> 
> I guess this is just a different method to achieve your second proposal (tabs always have seperate breadcumbs, and dedupe the url). I think that's much better.

Apps do always have a title and breadcrumb, so we're fine in that regard 👍 i was highlighting it in regards to dominik's suggestion of skipping `Settings` in dashboard settings. the equivalent for that in Apps would be to skip the app name which wouldn't work.

yes, i think what you're suggesting is equivalent to proposal #2

**Comment by ashharrison90** [comment]
@joshhunt @dprokop have updated the PR with the suggestion of having separate breadcrumbs. i don't love it, but i like it more than the existing behaviour so it's progress at least 😄 

on the ephemeral instance, the 2 examples where this is most obvious are:
- the tabs on dashboard settings
- the tabs under AWS -> dashboards, e.g. [here](https://ephemeral1511182189102ashharr.grafana-dev.net/a/grafana-aws-app/dashboard/ec2/region?var-datasource=grafanacloud-prom&var-region=$__all&var-allTags=&var-allTagsValues=)

lmk your thoughts 🙏

**Review by joshhunt — APPROVED** [review]
nice. i think this works well

**Pull request merged**